vpdq/cpp/hashing/matchTwoHash.cpp

```diff
@@ -2,82 +2,99 @@
+ *
+ * @param features Features to filter
+ * @param qualityTolerance Quality tolerance of comparing two hashes. If lower
+ * then it won't be included in the result
+ * @param verbose Print skipped hashes
+ *
+ * @return Feature vector without features with quality lower than
+ * qualityTolerance
+ */
+static std::vector<vpdq::hashing::vpdqFeature> filterFeatures(
+    const std::vector<vpdq::hashing::vpdqFeature>& features,
+    const int qualityTolerance,
+    const bool verbose) {
+  std::vector<vpdq::hashing::vpdqFeature> filteredHashes;
+  for (const auto& feature : features) {
+    if (feature.quality >= qualityTolerance) {
+      filteredHashes.push_back(feature);
     } else if (verbose) {
-      printf(
-          "Skipping Line %zu Skipping Target Hash: %s, because of low quality Target Hash: %d \n",
-          i,
-          tHashes[i].pdqHash.format().c_str(),
-          tHashes[i].quality);
+      auto index = &feature - &features[0];
+      std::cout << "Skipping Line " << index
+                << " Skipping Hash: " << feature.pdqHash.format()
+                << ", because of low quality: " << feature.quality << std::endl;
     }
   }
+  return filteredHashes;
+}
 
-  for (i = 0; i < queryFiltered.size(); i++) {
-    for (j = 0; j < targetFiltered.size(); j++) {
-      if (queryFiltered[i].pdqHash.hammingDistance(targetFiltered[j].pdqHash) <
-          distanceTolerance) {
-        qMatchCnt++;
-        if (verbose) {
-          printf(
-              "Query Hash: %s Target Hash: %s match \n",
-              queryFiltered[i].pdqHash.format().c_str(),
-              targetFiltered[j].pdqHash.format().c_str());
-        }
-        break;
-      }
-    }
-  }
-  for (i = 0; i < targetFiltered.size(); i++) {
-    for (j = 0; j < queryFiltered.size(); j++) {
-      if (targetFiltered[i].pdqHash.hammingDistance(queryFiltered[j].pdqHash) <
+/**
+ * @brief Get the number of matches between two feature vectors
+ *
+ * @param features1 Features to match
+ * @param features2 Features to match
+ * @param distanceTolerance Distance tolerance of considering a match. Lower is
+ * more similar.
+ * @param verbose Print features with matching hashes
+ *
+ * @return Number of matches
+ */
+static unsigned int findMatches(
+    const std::vector<vpdq::hashing::vpdqFeature>& features1,
+    const std::vector<vpdq::hashing::vpdqFeature>& features2,
+    const int distanceTolerance,
+    const bool verbose) {
+  unsigned int matchCnt = 0;
+  for (const auto& feature1 : features1) {
+    for (const auto& feature2 : features2) {
+      if (feature1.pdqHash.hammingDistance(feature2.pdqHash) <
           distanceTolerance) {
-        tMatchCnt++;
+        matchCnt++;
         if (verbose) {
-          printf(
-              "Query Hash: %s Target Hash: %s match \n",
-              queryFiltered[j].pdqHash.format().c_str(),
-              targetFiltered[i].pdqHash.format().c_str());
+          std::cout << "Query Hash: " << feature1.pdqHash.format()
```

Comment:
(pls ignore): In an alternate universe, we could consider switching to use whatever the GLOG has turned into, since libraries writing directly to screen is a little odd. Way out of scope for this guy, but something for us to consider in the future.